### ISSUE TITLE
ganeti-noded binds to 0.0.0.0 after re-adding a node in the cluster even if told otherwise

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#323]
Originally reported of Google Code with ID 267.

```
What software version are you running? Please provide the output of "gnt-
cluster --version" and "gnt-cluster version".

# gnt-cluster --version
gnt-cluster (ganeti v2.5.1) 2.5.1

# gnt-cluster version
Software version: 2.5.1
Internode protocol: 2050000
Configuration format: 2050000
OS api version: 20
Export interface: 0


<b>What distribution are you using?</b>

# lsb_release -rd
Description:	Ubuntu 10.04.4 LTS
Release:	10.04


<b>What steps will reproduce the problem?</b>
1. Use NODED_ARGS="-b 172.20.31.13"
2. Confirm noded binds to 172.20.31.13
3. Remove a node from cluster for maintenance
4. Add the node back in the cluster
5. Confirm the noded binds to 0.0.0.0

What is the expected output?

I would expect the noded to keep binding to the specified IP instead of 0.0.0.0.

<b>Please provide any additional information below.</b>

The node is configured to bind the noded to the node IP instead of 0.0.0.0 as shown :

root@node3:~# cat /etc/default/ganeti
# File managed by Puppet

# Default arguments for Ganeti daemons
NODED_ARGS="-b 172.20.31.13"
MASTERD_ARGS=""
RAPI_ARGS="-b 172.20.31.254"
CONFD_ARGS="-b 172.20.31.13"

Before re-adding the node to the cluster (node was removed for reboot):
root@node3:~# ps aux| grep python
root      2710  0.0  0.0  57840 16944 ?        SL   10:41   0:00 /usr/bin/python /usr/sbin/ganeti-noded -b 172.20.31.13
root      2780  0.0  0.0  58228  9720 ?        S    10:41   0:00 /usr/bin/python /usr/sbin/ganeti-confd -b 172.20.31.13
root      5766  0.0  0.0   7624   956 pts/0    S+   10:49   0:00 grep --color=auto python

root@node3:~# netstat -puant| grep python
tcp        0      0 172.20.31.13:1811       0.0.0.0:*               LISTEN      2710/python     
udp        0      0 172.20.31.13:1814       0.0.0.0:*                           2780/python 

Right after re-adding it to the cluster:

root@node3:~# ps aux| grep python
root      5872  0.2  0.0  57832 16936 ?        SL   10:49   0:00 /usr/bin/python /usr/sbin/ganeti-noded -b 172.20.31.13 -b 0.0.0.0
root      5995  1.0  0.0  58248  9680 ?        S    10:50   0:00 /usr/bin/python /usr/sbin/ganeti-confd -b 172.20.31.13
root      5997  0.0  0.0   7624   956 pts/0    S+   10:50   0:00 grep --color=auto python

root@node3:~# netstat -puant| grep python
tcp        0      0 0.0.0.0:1811            0.0.0.0:*               LISTEN      5872/python     
udp        0      0 172.20.31.13:1814       0.0.0.0:*                           5995/python     


Then, if the ganeti service is manually restarted, everything goes back to normal:

root@node3:~# ps aux| grep python
root      6934  0.3  0.0  58124 17228 ?        SL   10:50   0:00 /usr/bin/python /usr/sbin/ganeti-noded -b 172.20.31.13
root      6993  0.3  0.0  58540  9720 ?        S    10:50   0:00 /usr/bin/python /usr/sbin/ganeti-confd -b 172.20.31.13
root      6997  0.0  0.0   7624   952 pts/0    S+   10:50   0:00 grep --color=auto python

root@node3:~# netstat -puant| grep python
tcp        0      0 172.20.31.13:1811       0.0.0.0:*               LISTEN      6934/python     
udp        0      0 172.20.31.13:1814       0.0.0.0:*                           6993/python 
```
Originally added on 2012-09-12 21:52:54 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
The problematic function appears to be SetupNodeDaemon() from /usr/share/pyshared/ganeti/bootstrap.py:

def SetupNodeDaemon(cluster_name, node, ssh_key_check):
  ...

  family = ssconf.SimpleStore().GetPrimaryIPFamily()
  sshrunner = ssh.SshRunner(cluster_name,
                            ipv6=(family == netutils.IP6Address.family))

  bind_address = constants.IP4_ADDRESS_ANY
  if family == netutils.IP6Address.family:
    bind_address = constants.IP6_ADDRESS_ANY

  # set up inter-node password and certificate and restarts the node daemon
  # and then connect with ssh to set password and start ganeti-noded
  # note that all the below variables are sanitized at this point,
  # either by being constants or by the checks above
  sshrunner.CopyFileToNode(node, constants.NODED_CERT_FILE)
  sshrunner.CopyFileToNode(node, constants.RAPI_CERT_FILE)
  sshrunner.CopyFileToNode(node, constants.CONFD_HMAC_KEY)
  mycommand = ("%s stop-all; %s start %s -b %s" %
               (constants.DAEMON_UTIL, constants.DAEMON_UTIL, constants.NODED,
                utils.ShellQuote(bind_address)))
  ...
```
Originally added on 2012-09-12 21:53:12 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Just came across this myself yesterday while working on virtual clusters.
```
Originally added on 2012-09-14 09:14:18 +0000 UTC.

Changed State: Accepted

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit 93b19295f58c5419d9cfcdbb4eb5ef8e7bd0b83a
Author: Michael Hanselmann <hansmi@google.com>
Date:   Thu Sep 20 18:52:49 2012 +0200

    bootstrap.SetupNodeDaemon: Stop hardcoding bind address
    
    Commit b43dcc5a11 added support for IPv6. To have the node daemon bind
    to an IPv6 address on startup it changed the code to pass a static bind
    address (0.0.0.0 or ::0). If a user had configured another bind address
    using /etc/default/ganeti, that address would not be used until the
    daemons were restarted.
    
    By copying ssconf files before attempting to start the daemons we can
    ensure that the built-in IP family detection in daemon.py works.
    
    This fixes issue 267.
    
    Signed-off-by: Michael Hanselmann <hansmi@google.com>
    Reviewed-by: René Nussbaumer <rn@google.com>

```
Originally added on 2012-09-21 12:47:13 +0000 UTC.

Mentioned Issue with GoogleCode ID 267 has been migrated to GitHub with ID #323.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Thanks a lot Michael for looking into this. I just applied your patch with some modifications to apply on 2.5.1 and that works well. Many thanks.

The modified patch for 2.5.1 if anyone needs it.
```
Originally added on 2012-09-21 13:48:00 +0000 UTC.

Attachments: 
[ganeti-2.5.1-issue-267.patch](https://storage.googleapis.com/google-code-attachments/ganeti/issue-267/comment-4/ganeti-2.5.1-issue-267.patch)

**Comment by Ganeti-Issues-Migrator** [comment]
```
Marking as fixed, this will be released in 2.7.
```
Originally added on 2012-10-30 15:45:04 +0000 UTC.

Added Labels: Type-Defect
Changed State: Fixed
Added to Milestone: Release2.7

**Comment by Ganeti-Issues-Migrator** [comment]
```
Marking 2.7 bugs fixed, as 2.7.0~rc1 was released, and should be ready to be tested in production environments.
```
Originally added on 2013-05-09 05:42:48 +0000 UTC.

Changed State: Released
